src/backend/distributed/utils/shard_pruning.c

```diff
@@ -0,0 +1,1239 @@
+				 * used.
+				 */
+				instance->instance->isPartial = true;
+
+				context->pendingInstances = lappend(context->pendingInstances, instance);
+			}
+
+			return false;
+		}
+	}
+	else if (IsA(node, OpExpr))
+	{
+		OpExpr *operatorExpression = (OpExpr *) node;
+		PruningInstance *prune = context->currentPruningInstance;
+		Node *leftOperand = NULL;
+		Node *rightOperand = NULL;
+		Const *constantClause = NULL;
+		Var *varClause = NULL;
+
+		if (!prune->added)
+		{
+			context->pruningInstances = lappend(context->pruningInstances,
+												prune);
+			prune->added = true;
+		}
+
+		if (list_length(operatorExpression->args) == 2)
+		{
+			leftOperand = get_leftop((Expr *) operatorExpression);
+			rightOperand = get_rightop((Expr *) operatorExpression);
+
+			leftOperand = strip_implicit_coercions(leftOperand);
+			rightOperand = strip_implicit_coercions(rightOperand);
+
+			if (IsA(rightOperand, Const) && IsA(leftOperand, Var))
+			{
+				constantClause = (Const *) rightOperand;
+				varClause = (Var *) leftOperand;
+			}
+			else if (IsA(leftOperand, Const) && IsA(rightOperand, Var))
+			{
+				constantClause = (Const *) leftOperand;
+				varClause = (Var *) rightOperand;
+			}
+		}
+
+		if (constantClause && varClause && equal(varClause, context->partitionColumn))
+		{
+			List *btreeIntepretationList = NULL;
+			ListCell *btreeInterpretationCell = NULL;
+			bool matchedOp = false;
+
+			/*
+			 * Found a restriction on the partition column itself. Update the
+			 * current constraint with the new information.
+			 */
+
+			btreeIntepretationList =
+				get_op_btree_interpretation(operatorExpression->opno);
+			foreach(btreeInterpretationCell, btreeIntepretationList)
+			{
+				OpBtreeInterpretation *btreeIntepretation =
+					(OpBtreeInterpretation *) lfirst(btreeInterpretationCell);
+
+				switch (btreeIntepretation->strategy)
+				{
+					case BTLessStrategyNumber:
+					{
+						if (!prune->lessConsts ||
+							PerformValueCompare(&context->compareValueFunctionCall,
+												constantClause->constvalue,
+												prune->lessConsts->constvalue) < 0)
+						{
+							prune->lessConsts = constantClause;
+						}
+						matchedOp = true;
+					}
+					break;
+
+					case BTLessEqualStrategyNumber:
+					{
+						if (!prune->lessEqualConsts ||
+							PerformValueCompare(&context->compareValueFunctionCall,
+												constantClause->constvalue,
+												prune->lessEqualConsts->constvalue) < 0)
+						{
+							prune->lessEqualConsts = constantClause;
+						}
+						matchedOp = true;
+					}
+					break;
+
+					case BTEqualStrategyNumber:
+					{
+						if (!prune->equalConsts)
+						{
+							prune->equalConsts = constantClause;
+						}
+						else if (PerformValueCompare(&context->compareValueFunctionCall,
+													 constantClause->constvalue,
+													 prune->equalConsts->constvalue) != 0)
+						{
+							/* key can't be equal to two values */
+							prune->alwaysFalse = true;
```

Comment:
We're checking for comparator() != 0 above?  That precludes such a problem, right?